File: Binary_Tree/src/BST/BST.java
package BST;

public class BST implements BST_Interface {
	public BST_Node root;
	int size;

	public BST() {
		size = 0;
		root = null;
	}

	@Override
	// used for testing, please leave as is
	public BST_Node getRoot() {
		return root;
	}

	@Override
	public boolean insert(String s) {
		if(s==null)return false;
		if (root == null) {
			root = new BST_Node(s);
			size++;
			return true;
		}
		if (contains(s))
			return false;

		BST_Node insert_node = new BST_Node(s);
		BST_Node target_node = root;

		while (true) {
			int i = target_node.data.compareTo(s);
			if (i > 0) {
				if (target_node.left == null) {
					target_node.left = insert_node;
					break;
				} else {
					target_node = target_node.left;

				}

			}
			if (i < 0) {
				if (target_node.right == null) {
					target_node.right = insert_node;
					break;
				} else {
					target_node = target_node.right;

				}

			}

		}
		size++;
		return true;

	}

	@Override
	public boolean remove(String s) {
		if(s == null) return false;
		BST_Node node = findNode(s);
		BST_Node target_node = root;
		if (node == null)
			return false;
		
		//if the removed node is left to its parent
		boolean left= false;
		int i = target_node.data.compareTo(s);

		while (i!=0) {
			if (target_node.left == node | target_node.right == node)
				break;
			i = target_node.data.compareTo(s);
			if (i > 0)
				target_node = target_node.left;
			if (i < 0)
				target_node = target_node.right;
			
		}
		if (target_node.left == node) left = true;
		// leaf delete
		if (node.left == null & node.right == null) {
			if(node == root) {
				root= null;
			}else
			if(left) {
				target_node.left = null;
			}else {
				target_node.right = null;
			}
			size--;
		
		// one child delete
		} else if (node.left != null ^ node.right != null) {
			BST_Node child;
			if(node.left != null) {
				child = node.left;
			}else {
				child = node.right;
			}
			if(node == root) {
				root = child;
			}else
			if(left) {
				target_node.left = child;
			}else {
				target_node.right = child;
			}
			size--;
			// two child delete
		}else {
			BST_Node child = minNode(node.right);
			String tmp = child.data;
			remove(child.data);
			node.data = tmp;
		}
		
		
		
			
			return true;
	}

	private BST_Node minNode(BST_Node target_node) {
		while (target_node.left != null)
			target_node = target_node.left;
		return target_node;
	}

	public String findMin() {
		if(root == null) return null;
		return minNode(root).getData();
	}

	@Override
	public String findMax() {
		if(root == null) return null;
		return maxNode(root).getData();
	}

	private BST_Node maxNode(BST_Node target_node) {
		while (target_node.right != null)
			target_node = target_node.right;
		return target_node;
	}

	@Override
	public boolean empty() {
		// TODO Auto-generated method stub
		return size == 0;
	}

	@Override
	public boolean contains(String s) {
		if (findNode(s) == null)
			return false;
		return true;
	}

	private BST_Node findNode(String s) {
		if (size == 0)
			return null;
		BST_Node target_node = root;

		for (int i = target_node.data.compareTo(s); i != 0; i = target_node.data.compareTo(s)) {
			if (i > 0)
				target_node = target_node.left;
			if (i < 0)
				target_node = target_node.right;
			if (target_node == null)
				return null;
		}
		return target_node;

	}
	@Override

	public int size() {
		// TODO Auto-generated method stub
		return size;
	}
	private int dive(BST_Node node) {
		if(node == null) return 0;
		if(node.left==null & node.right == null) return 1;
		return 1+Integer.max(dive(node.left), dive(node.right));
	}
	@Override
	public int height() {

		
		return dive(root)-1;

	
	}

}

File: Pythonstuff/BST/BST.py
'''
Created on Feb 13, 2020

@author: yazid
'''
from BST_Node import BST_Node


class BST(object):
    
    root= None 
    size = int(0)
    


    def __init__(self):
        pass
    
    def findNode(self, s):
        if self.size == 0:
            return None
        tgt = self.root
        while tgt.data != s:
            if(s > tgt.data):
                tgt = tgt.right
            elif s < tgt.data:
                tgt = tgt.left
            if tgt == None:
                return None
        return tgt
        
    def contains(self, s):
        if self.findNode(s) == None:
            return False
        return True
    
    def insert(self, s):
        if s == None:
            return False
        if self.root == None:
            self.root = BST_Node(s)
            self.size += 1
            return True
        if self.contains(s):
            return False 
        ins = BST_Node(s)
        tgt = self.root
        while True:
            if s > tgt.data:
                if tgt.right == None:
                    tgt.right = ins
                    break
                else:
                    tgt = tgt.right
            else:
                if tgt.left == None:
                    tgt.left = ins
                    break
                else:
                    tgt = tgt.left
        self.size += 1
        return True
    def remove(self, s):
        if s == None :
            return False
        node = self.findNode(s)
        tgt = self.root
        left = False
        if not self.contains(s):
            return False
        while s != tgt.data:
            if tgt.left is node or tgt.right is node:
                break
            if s > tgt.data:
                tgt = tgt.right
            else:
                tgt = tgt.left
        if tgt.left == node:
            left = True
        #leaf del
        if node.left == None and node.right == None:
            if node is self.root:
                self.root = None
            else:
                if left:
                    tgt.left = None
                else:
                    tgt.right = None
        elif node.left != None and node.right != None:
            child = self.minNode(node.right)
            tmp = node.data
            self.remove(child.data)
            node.data = tmp 
            return True
        else:
            if node.left != None:
                child = node.left
            else:
                child = node.right
            if node is self.root:
                self.root = child
            else:
                if left:
                    tgt.left = child
                else:
                    tgt.right = child
        self.size -= 1
        return True
            
                
            
    def minNode(self, node):
        while node.left != None:
            node = node.left
        return node
    def findMin(self):
        return self.minNode(self.root).data
    def findMax(self):
        node = self.root
        while node.right != None:
            node = node.right
        return node.data
    
    def dive(self, node):
        if(node == None):
            return 0
        if node.right == None and node.left == None:
            return 1
        return 1 + max(self.dive(node.left), self.dive(node.right))
        
    def height(self):
        
        return self.dive(self.root)-1
     
    def printlvl(self, node, nlvl, lvl , height):
        space = " "* (-(1-2**(height-nlvl+1))-1)
        if (node == None):
            s = "  "
            rnode = None
            lnode = None
        else:
            s= node.data
            rnode = node.right
            lnode= node.left
        if lvl == 0:
            return space +("[%s]" % s )+space
        else:
            s1 =self.printlvl(lnode, nlvl, lvl -1, height)
            s2 =self.printlvl(rnode, nlvl, lvl -1, height)
            return s1 + s2
    
    
    def __str__(self, *args, **kwargs):
        for i in range(self.height()+1):
            print self.printlvl(self.root, i, i, self.height())
        return ""
                
                
        
            
        